internal/registry/registry.go

```diff
@@ -0,0 +1,241 @@
+	}
+
+	dtk.OSVersion, found, err = unstructured.NestedString(obj.Object, "RHEL_VERSION")
+	if !found || err != nil {
+		return nil, fmt.Errorf("failed to get RHEL_VERSION from %s, found %t: %w", driverToolkitJSONFile, found, err)
+	}
+	return dtk, nil
+}
+
+func (r *registry) getImageManifest(ctx context.Context, image string, po *ootov1alpha1.PullOptions) ([]byte, *repoPullConfig, error) {
+	var repo string
+
+	if hash := strings.Split(image, "@"); len(hash) > 1 {
+		repo = hash[0]
+	} else if tag := strings.Split(image, ":"); len(tag) > 1 {
+		repo = tag[0]
+	}
+
+	if repo == "" {
+		return nil, nil, fmt.Errorf("image url %s is not valid, does not contain hash or tag", image)
+	}
+
+	var options []crane.Option
+	if po != nil && po.Insecure {
+		options = append(options, crane.Insecure)
+	}
+	manifest, err := r.getManifestStreamFromImage(image, repo, options)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get manifest stream from image %s: %w", image, err)
+	}
+
+	repoConfig := repoPullConfig{repo: repo, authOptions: options}
+
+	return manifest, &repoConfig, nil
+}
+
+func (r *registry) getManifestStreamFromImage(image, repo string, options []crane.Option) ([]byte, error) {
+	manifest, err := crane.Manifest(image, options...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get crane manifest from image %s: %w", image, err)
+	}
+
+	release := unstructured.Unstructured{}
+	if err = json.Unmarshal(manifest, &release.Object); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal crane manifest: %w", err)
+	}
+
+	imageMediaType, mediaTypeFound, err := unstructured.NestedString(release.Object, "mediaType")
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalled manifests invalid format: %w", err)
+	}
+	if !mediaTypeFound {
+		return nil, fmt.Errorf("mediaType is missing from the image %s manifest", image)
+	}
+
+	if strings.Contains(imageMediaType, "manifest.list") {
+		archDigest, err := r.getImageDigestFromMultiImage(manifest)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get arch digets from multi arch image: %w", err)
+		}
+		// get the manifest stream for the image of the architecture
+		manifest, err = crane.Manifest(repo+"@"+archDigest, options...)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get crane manifest for the arch image: %w", err)
+		}
+	}
+	return manifest, nil
+}
+
+func (r *registry) getLayersDigestsFromManifestStream(manifestStream []byte) ([]string, error) {
+	manifest := v1.Manifest{}
+
+	if err := json.Unmarshal(manifestStream, &manifest); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal manifest stream: %w", err)
+	}
+
+	digests := make([]string, len(manifest.Layers))
+	for i, layer := range manifest.Layers {
+		digests[i] = layer.Digest.Algorithm + ":" + layer.Digest.Hex
+	}
+	return digests, nil
+}
+
+func (r *registry) getHeaderFromLayer(layer v1.Layer, headerName string) (*unstructured.Unstructured, error) {
```

Comment:
This one looks unit testable, WDYT?
Also, should we return a better / more concrete type as the header?